Extract Classes to a separate macro

cecton opened the pull request:
#### Description

This PR will extract the code handling the classes from the macro `html!` and put it to its own macro `classes!`.

The intention is to keep the super helpful macro for class composition but move it outside so people can create libraries for yew and allow their user to use this macro.

This is not possible at the moment because it is strongly coupled to the `html!` macro and only for base HTML tags (a, div, span, etc...) but not for custom component.

#### Checklist

<!-- For further details, please read CONTRIBUTING.md -->

- [x] I have run `cargo make pr-flow`
- [x] I have reviewed my own code
- [x] I have added tests
  <!-- If this is a bug fix, these tests will fail if the bug is present (to stop it from cropping up again) -->
  <!-- If this is a feature, my tests prove that the feature works -->

Related to #1588

cecton commented on yew/src/virtual_dom/vtag.rs:
```diff
@@ -821,38 +822,41 @@ mod tests {
+                    .iter()
+                    .find_map(|(k, v)| if k == "class" { Some(v) } else { None })
             };
         }
 
         if let VNode::VTag(vtag) = a {
-            assert!(!has_class!(vtag));
+            assert_eq!(get_class!(vtag), Some(""));
```
This test has been inverted because an empty `Classes` will always generated an empty `String`

cecton commented on yew/src/virtual_dom/vtag.rs:
```diff
@@ -1022,23 +1026,6 @@ mod tests {
         html! { <div><a data-val=Box::<u32>::default() /></div> };
     }
 
-    #[test]
-    fn it_does_not_set_empty_class_name() {
-        let scope = test_scope();
-        let parent = document().create_element("div").unwrap();
-
-        #[cfg(feature = "std_web")]
-        document().body().unwrap().append_child(&parent);
-        #[cfg(feature = "web_sys")]
-        document().body().unwrap().append_child(&parent).unwrap();
-
-        let mut elem = html! { <div class=""></div> };
-        elem.apply(&scope, &parent, NodeRef::default(), None);
-        let vtag = assert_vtag(&mut elem);
-        // test if the className has not been set
-        assert!(!vtag.reference.as_ref().unwrap().has_attribute("class"));
-    }
```
I removed the entire test here because it is not true anymore

cecton commented on yew/Cargo.toml:
```diff
@@ -141,6 +141,9 @@ agent = ["bincode"]
 yaml = ["serde_yaml"]
 msgpack = ["rmp-serde"]
 cbor = ["serde_cbor"]
+class_macro = [
+	"yew-macro/class_macro"
+]
```
I don't think it will be possible to make a feature flag neither having both behavior working at the same time. Well, we CAN but we need to duplicated most of the tests.

siku2 replied:
Definitely don't want a feature flag for this.
Feature flags are additive across the entire dependency tree of an app (up and down) so if a library activates this feature it can be quite problematic.

cecton replied:
Yeah I already deleted it because it was kinda bad.

cecton commented on yew-macro/src/html_classes/mod.rs:
```diff
@@ -0,0 +1,32 @@
+use syn::token::Comma;
+use syn::Expr;
+
+/// List of HTML classes.
+pub struct HtmlClasses(Punctuated::<Expr, Comma>);
+
+impl Parse for HtmlClasses {
+    fn parse(input: ParseStream) -> Result<Self> {
+        Ok(HtmlClasses(Punctuated::<Expr, Comma>::parse_terminated(input)?))
+    }
+}
+
+impl ToTokens for HtmlClasses {
+    fn to_tokens(&self, tokens: &mut TokenStream) {
+        let n = self.0.len();
+        let classes = self.0.iter();
+        let new_tokens = quote! {
+            let mut __yew_classes = ::yew::virtual_dom::Classes::with_capacity(#n);
+            #(__yew_classes.push(#classes);)*
+            __yew_classes
+        };
+
+        tokens.extend(quote! {{
+            #[allow(clippy::useless_conversion, unused_braces)]
+            #new_tokens
+        }});
+    }
+}
```
The implementation of `classes!` is more trivial than the classes in `html!`. Here I have a comma separated list of expressions that I add (extend) one by on to `Classes`. If we remove the handling of classes in `html!`, the diff of this PR will be negative because the original implementation has more code.

cecton commented:
Hello @siku2 :wave: 

Here is finally another implementation for the classes that we talk about.

Since it is not possible to have the behavior that we have on "class" attributes in `html!` for custom components, I thought it would be maybe a good idea to actually split `html!` and make the part that handles classes its own macro `classes!`

Overall it works well and would even reduce the code size (after cleanup). But there are a few caveats:

1. I had to adapt all the tests to use the new macro. This actually means that if we want to support both ways (`html!` handling classes AND a new `classes!` macro, we would need to duplicated most of the tests. Sad.
2. There is no way for the attribute `class` to stay unset like before. Since `classes` is now a string, if you pass an empty `Classes` (or an empty str/String), it will become `class=""` at the end. I think the only way to avoid that would be to make an exception in `html!` for the `class` attribute but obviously it would re-open the main reason why this PR exists: having a consistent behavior for attributes `class` for custom components and base components. (Or maybe we could say `class` is not a String but an `Option<String>`? Just a wild idea...)

Since yew is still in version 0.x.y I think it would be okay to make a breaking change to the 0.18. The code would only force people to adapt by prefixing all their classes by `classes!` like this:

```rust
<div class=("a", "b") />
```

To

```rust
<div class=classes!("a", "b") />
```

Overall I think it's beneficial and it is worth the change (because the inconsistency bothers me) but you might have a different opinion. What do you think?

cecton commented on yew-macro/tests/macro/html-tag-fail.rs:
```diff
@@ -56,7 +56,6 @@ fn compile_fail() {
     html! { <a href?="href" /> };
     html! { <a href?=Some(NotToString) /> };
     html! { <input checked?=Some(false) /> };
-    html! { <input class?=() /> };
```
It did become possible now! Maybe there is something to do that would make it magically forget empty `Classes`

siku2 replied:
This was deliberately made impossible because `Option<T>` was already supported for `Classes` and it would completely break the syntax.
With these changes it makes sense for it to be possible now :+1:

cecton commented:
I don't pretend to have always excellent ideas :D don't hesitate to reject!

siku2 commented:
This isn't quite what I had in mind when I said "discussion" :upside_down_face:. My weak little heart wouldn't be able to reject this after you put in all this work...

I do like the idea of a `classes!` macro and I'm also generally in favour of making `class`'s special treatment more explicit. With that being said, there are a few things I need to bring up.

The biggest issue is that this is a super duper mega breaking change. This will force basically every single Yew app to update and it's not always as simple as adding `classes!` as a prefix.
Such a change certainly isn't inherently bad but I feel like there has to be a very good reason for it.
In this case you say it's:
> having a consistent behavior for attributes `class` for custom components and base components

The problem is that there isn't even a consistent behaviour across components.
Components could very well have `class: int` as a prop if they wanted to.
What I'm trying to get at is that components, right now, are inherently different from tags and I don't think it's healthy to pretend they're not.
That doesn't mean that I don't want for components to have access to the same convenience as tags do. I just don't think it's sensible to remove features from tags just because components don't have them.

Here are two approaches I'm comfortable with:

1. Make `classes!` implicit for tags so that `<div class="test" />` is like `<div class=classes!("test") />`. The `classes!` macro can then be used for components.
2. Turn `class` into a keyword for components so that `<MyComp class="test" />` works just like `<div class="test" />`


Again, the purist in me is totally in favour of moving classes out of `html!` into another macro but I just don't think we can justify it as it stands.

cecton commented:
> This isn't quite what I had in mind when I said "discussion" upside_down_face. My weak little heart wouldn't be able to reject this after you put in all this work...

Oops! :sweat: I didn't mean to force your hand. I felt like I needed to test a bit more my idea to see if it is even possible and how it translates. Really I didn't spend that much time on it. This is fine!

> it's not always as simple as adding `classes!` as a prefix.

There are 2 things but nothing more:
1. adding `classes!` in front
2. `class=` might generated attributes with empty string

>  What I'm trying to get at is that components, right now, are inherently different from tags and I don't think it's healthy to pretend they're not.

I see your point. `className` has always been a special case in React too but less of a special case because it really is just a `string` there.

My point is only that it will make things more consistent in general. Not perfect.

>  That doesn't mean that I don't want for components to have access to the same convenience as tags do. I just don't think it's sensible to remove features from tags just because components don't have them.

This is a good point. It's not exactly removed either...

> Turn `class` into a keyword for components. Much like the `children` field this will force the `class` field to be of type `Classes` but it would mean that things like `<MyComp class=("something", "else") />` just work.

That is very bad for me. I mean the idea is good and well intended but custom components might accept multiple class props that are applied on different HTML tags (or even other props) in their own component. [Here](https://github.com/palantir/blueprint/blob/develop/packages/core/src/components/collapsible-list/collapsibleList.tsx#L58) is an example with blueprint having 2 classes property: visibleItemClassName and className (inferred from IProps). For me this is a no go.

> Make `classes!` implicit for tags so that `<div class="test" />` is like `<div class=classes!("test") />`. The `classes!` macro can then be used for components. Like #1599 but for `Classes`.

I wanted to answer first before re-modifying my PR again so I don't jump from one thing to another and have a hard time to explain what happens in my head... but I have an idea :grin: similar to what you proposed.

Since `class=()` is just parse like a `TupleExpr` I guess I can make a `Punctuated<Expr, Comma>` out of it and make it transparently work like it use to without duplicated code.

The only difference that would still remain is that `class=()` will still translate to `class=""`. But I believe simply adding a `?` here like this: `class?=()` might help.

There might also be a point about treating any attribute that uses the `(...)` syntax as class-like. But I'm not sure if it is a "good" thing that we should allow. \<insert meme Your-scientists-were-so-preoccupied-with-whether-or-not-they-could,-they-didn’t-stop-to-think-if-they-should\>

BTW, almost entirely not related. I was wondering why we use `indexmap` then I realized HashSet does not keep order so I asked DDG (not google) and this is what I found:

> The order in which the attributes are overwritten is not determined by the order the classes are defined in the class attribute, but instead where they appear in the CSS.

 -- Source: https://stackoverflow.com/questions/1321692/how-to-specify-the-order-of-css-classes

cecton commented on yew/src/virtual_dom/vtag.rs:
```diff
@@ -1090,26 +1075,26 @@ mod tests {
-        assert_class(html! { <p class=Some(string_ref) /> }, "some-class");
-        assert_class(html! { <p class=&Some(string.clone()) /> }, "some-class");
-        assert_class(html! { <p class=&Some(string_ref) /> }, "some-class");
+        assert_class(html! { <p class=(Some(str)) /> }, "some-class");
+        assert_class(html! { <p class=(Some(string.clone())) /> }, "some-class");
+        assert_class(html! { <p class=(Some(string_ref)) /> }, "some-class");
+        assert_class(html! { <p class=(&Some(string.clone())) /> }, "some-class");
+        assert_class(html! { <p class=(&Some(string_ref)) /> }, "some-class");
```
This can now be written `class?=Some(string)`.

I won't modify this test because it is suppose to test with the parenthesis but obviously it works too since class is just a normal attribute here.

cecton commented on yew-macro/tests/macro/html-tag-fail.stderr:
```diff
@@ -256,7 +250,7 @@ error[E0277]: the trait bound `std::borrow::Cow<'static, str>: std::convert::Fro
              <std::borrow::Cow<'a, [T]> as std::convert::From<std::vec::Vec<T>>>
              <std::borrow::Cow<'a, std::ffi::CStr> as std::convert::From<&'a std::ffi::CStr>>
-           and 11 others
+           and 12 others
```
I have no idea why this has increased :interrobang:

cecton commented on yew/src/virtual_dom/vtag.rs:
```diff
@@ -736,11 +736,10 @@ mod tests {
     fn supports_multiple_classes_string() {
         let a = html! {
-            <div class="class-1 class-2 class-3"></div>
+            <div class=("class-1 class-2 class-3")></div>
```
This won't split without parenthesis anymore but both syntax are actually working

cecton commented on yew/src/virtual_dom/vtag.rs:
```diff
@@ -821,38 +820,41 @@ mod tests {
     #[test]
     fn filter_empty_string_classes() {
-        let a = html! { <div class=vec![""]></div> };
+        let a = html! { <div class=(vec![""])></div> };
```
This syntax is now invalid without the parenthesis

cecton commented on yew/src/virtual_dom/vtag.rs:
```diff
@@ -821,11 +820,11 @@ mod tests {
-        let c = html! { <div class=""></div> };
+        let a = html! { <div class?=(vec![""])></div> };
+        let b = html! { <div class?=("", "")></div> };
+        let c = html! { <div class?=("")></div> };
         let d_arr = [""];
-        let d = html! { <div class=&d_arr[..]></div> };
+        let d = html! { <div class?=(&d_arr[..])></div> };
```
This now works \o/ But you need to specific that class is optional and you must use the parenthesis

cecton commented on yew/src/virtual_dom/vtag.rs:
```diff
@@ -1032,7 +1031,7 @@ mod tests {
         document().body().unwrap().append_child(&parent).unwrap();
 
-        let mut elem = html! { <div class=""></div> };
+        let mut elem = html! { <div class?=("")></div> };
```
I have put back this test. Obviously it works also for `class?=""` since class is a String. But it also works when using `()`

siku2 replied:
Just so you know, the question mark syntax is supposed to be reworked before 0.18 (see #1550).

*If you actually prefer the question mark syntax over the more implicit one proposed in the referenced issue then please voice your opinion there (I'm looking for excuses not to change it :P)*

cecton commented on yew-macro/src/html_tree/html_tag/mod.rs:
```diff
@@ -194,15 +194,44 @@ impl ToTokens for HtmlTag {
-                        quote! {
-                            ::yew::virtual_dom::PositionalAttr(#key, #sr)
+                    match value {
+                        Expr::Paren(ExprParen { expr, .. }) if label.name == "class" => {
```
if you remove `if label.name == "class"`, the `()` syntax can be used everywhere.

This would be valid for example:

```rust
<div style=("background-color: blue;", "color: green;") />
```

But this too:

```rust
<a href=() />
```
(translates to empty string)

It might be good. Idk. Let me know!

cecton commented:
> I have an idea grin similar to what you proposed.

It worked. This is now ready for review.

I remove the empty file `yew/src/prelude.rs`. Was it there for a purpose?

cecton commented on yew/src/virtual_dom/vtag.rs:
```diff
@@ -791,7 +790,7 @@ mod tests {
         classes.push("class-2");
         let a = html! {
-            <div class=classes></div>
+            <div class=(classes)></div>
```
This must now be written with `()`

cecton commented:
> There might also be a point about treating any attribute that uses the `(...)` syntax as class-like. But I'm not sure if it is a "good" thing that we should allow. <insert meme Your-scientists-were-so-preoccupied-with-whether-or-not-they-could,-they-didn’t-stop-to-think-if-they-should>

I have been thinking about this and I kinda had a change of heart about it.

One reason is because the tokens `(` and `)` in the macro are somehow already dedicated to classes. At the moment it is dedicated only to the attribute `class` but you can't build anything else with it, so it is somehow dedicated to classes already. `()` are often use to group things but in yew (and JSX) we use `{}` for that purpose (`href={self.props.href}`). So I think: unless we want to add more meaning to these tokens, it is fine to dedicate them to classes.

The other reason is the consistency: it is kinda nice to have the same access to the features on all attributes. I'm not talking about custom components, I mean in general: it feels less surprising than having `<foo class=(baz)>` being OK while `<foo bar=(baz)>` being NOT OK. (See hwat I mean?)

So I would be in favor to drop the `if attribute == "class"` in the match arms.

siku2 commented:
> There are 2 things but nothing more:
> 
>     1. adding `classes!` in front
> 
>     2. `class=` might generated attributes with empty string

There's also the case where class isn't already a tuple but you're right. For some reason it seemed much more complicated in my head than it actually is.

The second point is a big problem but I haven't mentioned it because it's something that can be fixed so let's just ignore that for now.

<br>

> I mean the idea is good and well intended but custom components might accept multiple class props that are applied on different HTML tags (or even other components) in their own component. [Here](https://github.com/palantir/blueprint/blob/develop/packages/core/src/components/collapsible-list/collapsibleList.tsx#L58) is an example with blueprint having 2 classes property: visibleItemClassName and className (inferred from IProps)

That's a very good point, thanks for bringing it up. I see that the `class` attribute isn't really as special as `children` is. It's `Classes` that's special.

<br>

> BTW, almost entirely not related. I was wondering why we use `indexmap` then I realized HashSet does not keep order so I asked DDG (not google) and this is what I found:

Ha, interesting find. That's not the reason we're using `IndexMap`.
We want the classes to have a consistent ordering when diffing the virtual DOM. Because `HashMap`s don't guarantee any particular order `class1.to_string() == class2.to_string()` doesn't necessarily hold true even when `class1 == class2`.

Using `IndexMap` also has the drawback though. The values are ordered in insertion ordered so two `Classes` containing the same strings might not create the same output depending on what order they were originally added.
Ideally we should replace this to have the best of both worlds (e.g. by sorting the classes alphabetically) but that's not the burning issue right now.

<br>

> I wanted to answer first before re-modifying my PR again so I don't jump from one thing to another and have a hard time to explain what happens in my head... but I have an idea grin similar to what you proposed.

Well... You went ahead and jumped the gun again, didn't you? If it's easier for you to express your ideas this way then by all means, go for it! I just want you to know that I'm not expecting you to do it and I'll try not to feel heartbroken about dismissing the work.

<br>

So anyway, let me finally get to the changes you've made... and well... I'm sorry but I feel like this is a much inferior version of the `classes!` macro idea.

- This new syntax is nowhere near as explicit as `classes!` (which also works outside of the `html!` macro) was.
- The syntax isn't really Rust or HTML. We try to keep the "fictional" syntax to a minimum. Sure, tuples are a thing but they don't work like this, so the similarity is nothing but a false friend. 
- This also isn't any less of a breaking change.
- Users might start using this syntax outside of places where `Classes` is appropriate. It will work because `Classes` implements `ToString` so it can be used for any attribute and without the explicit `classes!()` syntax one might easily believe that this is just a handy way of joining strings with a space. Needless to say that this can lead to unexpected issues. 


Now that I've made myself feel all terrible again, let me propose a new approach based on the first approach from prior:

The idea is that we add `classes!` *and* keep the current special treatment for tags but we deprecate it so it can be removed somewhere down the line.
I realise that Yew isn't stable yet and breaking changes should be expected, but I'm uncomfortable making such a big breaking change when it can't justify itself through convenience or potential use cases.
Let's be honest, this is mostly a change for internal consistency which, personally, I can't get enough of :) but I can imagine it being a hard sell for most users.
By deprecating the old way first we can establish a much better transition to the new macro which doesn't immediately break any project that tries to update.
The last thing I want is for Yew apps to stay on 0.17 just because they don't want to have to update every `class` attribute.

What do you think, does that sound reasonable at all?
There are still many aspects about the macro itself (and the backing `Classes` implementation) I'd like to discuss, but I really want to nail down the concept first.

cecton commented:
> This new syntax is nowhere near as explicit as `classes!` (which also works outside of the `html!` macro) was.

(It's actually doing both: the `()` AND the `classes!()` macro! yeah yeah)

> * We try to keep the "fictional" syntax to a minimum.

:+1: 

> This also isn't any less of a breaking change.

:-1: A lot less tests are broken by these latest changes. Might not be a great measure for breaking changes though... :thinking: 

> * Users might start using this syntax outside of places where `Classes` is appropriate. It will work because `Classes` implements `ToString` so it can be used for any attribute and without the explicit `classes!()` syntax one might easily believe that this is just a handy way of joining strings with a space. Needless to say that this can lead to unexpected issues.

:+1: Agreed. I'm not a big fan either for that reason. I'm looking at your wisdom, I'm sure you will know what's best, I'm imagining new propositions.

> Now that I've made myself feel all terrible again,

Really you shouldn't be :heart: 

> The idea is that we add `classes!` _and_ keep the current special treatment for tags but we deprecate it so it can be removed somewhere down the line.

This sounds very reasonable.

> What do you think, does that sound reasonable at all?

Ok!

jstarry commented:
> Ideally we should replace this to have the best of both worlds (e.g. by sorting the classes alphabetically) but that's not the burning issue right now.

There are some css frameworks that rely on class order and even expect duplicate classes I believe. Context here: https://github.com/yewstack/yew/issues/393

cecton commented:
> The idea is that we add `classes!` _and_ keep the current special treatment for tags but we deprecate it so it can be removed somewhere down the line.

@siku2 Done! It was fast actually... How do I add a deprecation warning or something?

siku2 commented:
@cecton That's the hard part. On stable it's not yet possible to emit a warning directly so we're gonna have to get creative.
One thing you could try is generating code that calls a deprecated method:

```rust
#[deprecated(note="blabla deprecated blabla use classes! macro")]
fn deprecated() {}

if false {
  // it's important that these tokens have the correct span (use `quote_spanned`) for the error to be visible.
  deprecated()
}
```

cecton commented on yew-macro/tests/macro/html-tag-fail.stderr:
```diff
@@ -148,17 +148,19 @@ error: the `checked` attribute does not support being used as an optional attrib
 
-error: the `ref` attribute does not support being used as an optional attribute
-  --> $DIR/html-tag-fail.rs:60:20
+warning: use of deprecated item 'compile_fail::deprecated_use_of_class': blabla deprecated blabla use classes! macro
```
This message is :ok_hand: purrfect :joy_cat:

siku2 commented:
Apologies for the long wait time.

The biggest problem now (apart from the many nits I won't be able to hold back :pray: and some other chores) is that empty `classes!()` still adds an empty attribute to the DOM.
I think the best way to solve this is with optional attributes but because of that, this is currently blocked by #1550. I promise I'll get that done soon. Only reason I haven't is because I've been working on #1599 and I didn't want to deal with the merge conflicts :see_no_evil:

There's still a problem though. `classes!` should always return `Classes`, not `Option<Classes>`. We need to find a way to do the conversion from `Classes -> Option<String>` which is not something we can't do with the `ToString` trait.
But we'll cross that bridge when we get there.

cecton commented:
Ok ok no hurry! 😅  It's not a bug fix, I'm not awaiting it.

Let me know if there is something I could help on

siku2 commented:
The changes to optional attributes is taking a lot longer than I expected. So  we might as well go ahead with this because I would like to have this included in 0.18 as well. With my intended changes in #1637 it will be easy to take care of the empty classes handling there so we don't need to worry about it here.

There are two things that still need to be done here:

- Add dedicated macro tests for `classes!` (see the `props!` macro tests for reference). It doesn't need to be much, I mainly want to have tests for the error messages and their spans.
- Update the documentation for "Classes". We should probably move it from "Elements" to somewhere else too. Maybe even give it its own "concept" page. Let me know if you have an idea what it should look like.

siku2 commented on yew-macro/src/props/element.rs:
```diff
@@ -1,19 +1,24 @@
 pub enum ClassesForm {
-    Tuple(Vec<Expr>),
-    Single(Box<Expr>),
+    Tuple(Span, Vec<Expr>),
```
We can make this much prettier by passing `ExprTuple` directly instead of `Vec<Expr>`. That way we don't need to have a separate field for the span.

cecton replied:
That's genius!

Fixed in e439e6e

siku2 commented on yew-macro/src/html_classes/mod.rs:
```diff
@@ -0,0 +1,34 @@
+    }
+}
+
+impl ToTokens for HtmlClasses {
```
To give the `classes!` macro a bit more of an edge (and justify it being a proc macro instead of a declarative macro) I would like is to add special treatment for string literals.
Most `classes!` invocations will probably involve at least one string literal and that way we can optimize them:
- String literals containing a space should error with the suggestion of splitting them apart (`classes!("hello world")` should be written as `classes!("hello", "world")`, because with the latter it's much clearer what's going on).
- They should use an unsafe `unchecked_push` method that adds a class to the `Classes` value without having to split it up first. The compiler *might* be smart enough to do this already, but by doing it explicitly we can be sure it's happening and hopefully even cut down on compile times. 

Do you think you could do that?

cecton replied:
Fixed in e748001 (ensure not space in class name)

Fixed in c30ee6c (use `unchecked_push()` for string literals)

cecton replied:
```
error: string literals should not contain spaces: please use two separate literals
   --> yew/src/virtual_dom/classes.rs:303:44
    |
303 |             <div class=classes!("class-1", "class-1 class-2")></div>
    |                                            ^^^^^^^^^^^^^^^^^

error: string literals should not contain spaces: please use two separate literals
   --> yew/src/virtual_dom/classes.rs:318:33
    |
318 |             <div class=classes!("class-1 class-2 class-3")></div>
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^

error: string literals should not contain spaces: please use two separate literals
   --> yew/src/virtual_dom/classes.rs:321:33
    |
321 |             <div class=classes!("class-2 class-3 class-1")></div>
    |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^
```

I guess it worked :joy: Now I need to fix the other tests

siku2 reviewed (commented):
And here's a review of the current code. Apart from my feature suggestion everything is pretty minor stuff.

I have a few more things in mind regarding the `Classes` struct itself, but it can wait to keep the scope of this one reasonable.

cecton commented on yew-macro/src/html_classes/mod.rs:
```diff
@@ -0,0 +1,34 @@
+use quote::{quote, ToTokens};
+use syn::parse::{Parse, ParseStream, Result};
+use syn::punctuated::Punctuated;
+use syn::token::Comma;
```
Fixed in e4a70a5

cecton commented on yew-macro/src/html_classes/mod.rs:
```diff
@@ -0,0 +1,34 @@
+    fn parse(input: ParseStream) -> Result<Self> {
+        Ok(HtmlClasses(Punctuated::<Expr, Comma>::parse_terminated(
+            input,
+        )?))
```
Fixed in 823befe

cecton commented on yew-macro/src/html_classes/mod.rs:
```diff
@@ -0,0 +1,34 @@
+use syn::Expr;
+
+/// List of HTML classes.
+pub struct HtmlClasses(Punctuated<Expr, Comma>);
```
Fixed in 808ec09

cecton commented on yew-macro/src/html_classes/mod.rs:
```diff
@@ -0,0 +1,34 @@
+        };
+
+        tokens.extend(quote! {{
+            #[allow(clippy::useless_conversion, unused_braces)]
```
Fixed in af582f4

cecton replied:
I'm not sure about this.

```rust
    fn to_tokens(&self, tokens: &mut TokenStream) {
        let n = self.0.len();
        let classes = self.0.iter();
        let new_tokens = quote! {
            let mut __yew_classes = ::yew::virtual_dom::Classes::with_capacity(#n);
            #(__yew_classes.push(#classes);)*
            __yew_classes
        };

        tokens.extend(quote! {{
            #new_tokens
        }});
    }
```

I think if classes is empty, it should complain that `__yew_classes` doesn't need to be mutable. But I couldn't trigger the warning.

siku2 replied:
It probably does emit that warning but Clippy is actually pretty smart, it ignores the error if it doesn't originate from user code tokens. More specifically, because you're using `quote!` here all the tokens (apart from `classes`) default to the call site span which Clippy knows to ignore.
Sometimes this doesn't work because we're using `quote_spanned!` which attaches a span to all the tokens (which is, of course, most often the span from an input token which makes Clippy think that the user is at fault).

cecton replied:
Good to know... thanks!

cecton commented on yew-macro/src/props/element.rs:
```diff
@@ -1,19 +1,24 @@
-    Tuple(Vec<Expr>),
-    Single(Box<Expr>),
+    Tuple(Span, Vec<Expr>),
+    Single(Span, Box<Expr>),
```
Fixed in ea02f37

cecton commented on yew-macro/src/html_tree/html_element.rs:
```diff
@@ -233,41 +233,45 @@ impl ToTokens for HtmlElement {
                     } else {
                         let sr = lit.stringify();
-                        Some(quote! {
+                        Some(quote_spanned! { *span => {
```
Hopefully fixed in 4adbf6b

cecton commented on yew-macro/src/lib.rs:
```diff
@@ -119,3 +121,9 @@ pub fn props(input: TokenStream) -> TokenStream {
+
+#[proc_macro]
+pub fn classes(input: TokenStream) -> TokenStream {
+    let classes = parse_macro_input!(input as HtmlClasses);
```
Fixed in cb94ee8

cecton commented on yew/src/lib.rs:
```diff
@@ -108,7 +108,7 @@ extern crate self as yew;
 /// [`html_nested!`]: ./macro.html_nested.html
 /// [Yew Docs]: https://yew.rs/docs/en/concepts/html/
-pub use yew_macro::html;
+pub use yew_macro::{classes, html};
```
Why didn't I think of that before! :confounded: 

Fixed in 9092665

cecton commented:
> Add dedicated macro tests for `classes!` (see the `props!` macro tests for reference). It doesn't need to be much, I mainly want to have tests for the error messages and their spans.

Done in af582f4

cecton commented on yew/src/virtual_dom/vtag.rs:
```diff
@@ -561,7 +561,7 @@ where
 mod tests {
     use super::*;
-    use crate::html;
+    use crate::{classes, html};
```
Fixed in 84d3aa4

siku2 commented:
How do you feel about moving the `classes` module over to `html`? It no longer has anything to do with the virtual dom and the `html` contains these sorts of things.

cecton commented:
> Update the documentation for "Classes". We should probably move it from "Elements" to somewhere else too. Maybe even give it its own "concept" page. Let me know if you have an idea what it should look like.

Check this out b0c3558

cecton commented:
>  How do you feel about moving the `classes` module over to `html`? It no longer has anything to do with the virtual dom and the `html` contains these sorts of things.

I think it's a very good point

cecton commented:
@siku2 This test fails to compile for some reason: https://github.com/yewstack/yew/pull/1601/checks?check_run_id=1370661529

:thinking: I have no idea why... can you help me?

cecton commented:
> @siku2 This test fails to compile for some reason: https://github.com/yewstack/yew/pull/1601/checks?check_run_id=1370661529
> 
> thinking I have no idea why... can you help me?

Nevermind I found the issue

siku2 commented on yew-macro/src/html_tree/html_element.rs:
```diff
@@ -234,39 +234,53 @@ impl ToTokens for HtmlElement {
                         let sr = lit.stringify();
-                        Some(quote! {
-                            #vtag.__macro_push_attribute("class", #sr);
+                        Some(quote_spanned! {span=>
+                            {
+                                #vtag.__macro_push_attribute("class", #sr);
+                            }
                         })
```
I'm sorry if I caused a bit of confusion with my previous comment, but this doesn't need to be in its own expression block and it also doesn't need to use `quote_spanned!` because `#sr` already has a span attached to it.
```suggestion
                       Some(quote! {
                            #vtag.__macro_push_attribute("class", #sr);
                       })
```

You see, `quote_spanned!` doesn't overwrite the spans of the tokens you insert like `#this`, it only sets a span for the new tokens. In this case that would be `.__macro_push_attribute("class", _);`.
There's no way that an error should be associated with these tokens in the first place (unless it's our fault like if the method doesn't exist). You might think that it wouldn't hurt to attach a span to these anyway, but as I mentioned previously that will "turn on" Clippy which means we suddenly have to worry about that.
In the other snippets you modified we might start seeing warnings like `"__yew_classes" does not need to be mutable` because it now has a span attached to it.

This comment applies to the snippet above and below as well. One thing you do have to watch out for is the deprecation warning, which, as we know now, needs to have a span attached to it for it to be shown to the user.
I would handle that like this (added some comments for clarity):
```rust
Some(ClassesForm::Tuple(classes)) => {
    let classes = classes.elems.iter().collect::<Vec<_>>();
    let n = classes.len();
    let sr = stringify::stringify_at_runtime(quote! { __yew_classes });

    // notice the use of quote_spanned! here so that Clippy thinks `deprecated_use_of_class()` belongs to the input tokens.
    // I also moved the function definition in here because then we don't need to worry about the dead code warning
    // Technically we could also define the function using `quote!` instead of `quote_spanned!` to avoid the issue
    let deprecation_warning = quote_spanned! {classes.span()=>
        #[deprecated(
            note = "the use of `(...)` with the attribute `class` is deprecated and will be removed in version 0.19. Use the `classes!` macro instead."
        )]
        fn deprecated_use_of_class() {}

        if false {
            deprecated_use_of_class();
        };
    };

    // now we don't need to use `quote_spanned!` here which means we don't need to worry about Clippy warnings for the code we write in here
    Some(quote! {
        let mut __yew_classes = ::yew::virtual_dom::Classes::with_capacity(#n);
        #(__yew_classes.push(#classes);)*

        // insert the `deprecation_warning` here. It still has the span we attached above, but everything else here doesn't
        #deprecation_warning

        if !__yew_classes.is_empty() {
            #vtag.__macro_push_attribute("class", #sr);
        } else {
            #vtag.__macro_push_attribute_placeholder("class");
        };
    })
}
```

cecton replied:
Thanks a lot for all the explanation!! :heart:

siku2 commented on yew-macro/tests/macro/classes-pass.rs:
```diff
@@ -0,0 +1,25 @@
+use yew::prelude::*;
+
+fn compile_pass() {
+    classes!("one", "two");
+    classes!("one");
+    classes!();
+
+    classes!(vec!["one", "two"]);
+    classes!(vec!["one"]);
+    classes!(Vec::<&'static str>::new());
+    classes!(vec!["one"], vec!["two"]);
+
+    let some = Some("one");
+    let none: Option<&'static str> = None;
+    classes!(some, none);
+
+    let string = "one".to_string();
+    classes!(string.clone());
+    classes!(string, "two", vec!["three"]);
+
+    let array = ["one", "two"];
+    classes!(&array[..]);
+}
```
This is where the `#![no_implicit_prelude]` attribute comes in. It helps us make sure that the macro doesn't accidentally use an item from the local scope. In order for it to be effective we also have to avoid bringing anything into scope. That makes things slightly more verbose, but proper macro hygiene is more important.

```suggestion
#![no_implicit_prelude]

fn compile_pass() {
    // multiple literals
    ::yew::classes!("one", "two");
    // single literal
    ::yew::classes!("one");
    // empty
    ::yew::classes!();

    // multiple expressions
    ::yew::classes!(::std::vec!["one"], ::std::vec!["two"]);
    // single expression
    ::yew::classes!(::std::vec!["one", "two"]);

    // optional classes
    ::yew::classes!(
        ::std::option::Option::Some("one"),
        ::std::option::Option::None::<&'static str>
    );

    // mixed types
    ::yew::classes!("one".to_owned(), "two", ::std::vec!["three"]);
}
```

I also...
- added comments to explain what each case is testing for
- removed some duplicate cases (as in, they weren't testing any specific feature of the macro). We only need to test the macro here and because it heavily relies on `Classes` we can just that separately

siku2 commented on yew-macro/tests/macro_test.rs:
```diff
@@ -23,6 +23,9 @@ fn tests() {
     t.pass("tests/macro/html-element-pass.rs");
     t.compile_fail("tests/macro/html-element-fail.rs");
+
+    t.pass("tests/macro/classes-pass.rs");
+    t.compile_fail("tests/macro/classes-fail.rs");
```
Please create a new file for the `classes!` macro tests and also move the test files into a dedicated folder (like the `props!` macro).

What's also pretty cool is that you can use a wildcard to select the tests:
```suggestion
    t.pass("tests/classes_macro/*-pass.rs");
    t.compile_fail("tests/classes_macro/*-fail.rs");
```

siku2 commented on yew/src/html/classes.rs:
```diff
@@ -0,0 +1,641 @@
+    /// If the provided class has already been added, this method will ignore it.
+    ///
+    /// This method won't check if there are multiple classes in the input string.
+    pub fn unchecked_push<T: Into<Cow<'static, str>>>(&mut self, class: T) {
```
This method should be unsafe because it requires the user to manually uphold an invariant.
```suggestion
    pub unsafe fn unchecked_push<T: Into<Cow<'static, str>>>(&mut self, class: T) {
```

cecton replied:
I'm not sure I understand :thinking:

siku2 replied:
Why this should be `unsafe`? It's for the same reason that something like `String::from_utf8_unchecked` is unsafe. It's not because the function itself can cause memory corruption or a crash, but because the function can be used to accidentally break an invariant. In the case of `String` this is that all strings need to be valid UTF8 and `from_utf8_unchecked` doesn't have a check in place to ensure that this is true.
In the case of `Classes` the invariant is that it's a list of separate classes. We enforce this by always splitting strings on whitespace, right? But this function doesn't do that and so it can be used to break this invariant. To make this clear to the user (like a little nudge to read the documentation) we mark it as unsafe.

cecton replied:
Makes sense. Thanks!! :heart:

siku2 commented on yew/Cargo.toml:
```diff
@@ -120,6 +120,7 @@ trybuild = "1.0"
 rustversion = "1.0"
 rmp-serde = "0.14.0"
 bincode = "1"
+boolinator = "2.4.0"
```
I don't think there's anything we need `boolinator` for to test, is there? Let's not add a dependency just for that...

cecton replied:
```suggestion
```

siku2 commented on yew/src/lib.rs:
````diff
@@ -98,6 +98,70 @@
+///
+/// ```
+/// # use yew::prelude::*;
+/// use boolinator::Boolinator;
+///
+/// struct MyComponent {
+///     props: MyComponentProps,
+/// }
+///
+/// #[derive(Clone, Properties)]
+/// struct MyComponentProps {
+///     fill: bool,
+///     class: Option<String>,
+///     children: html::Children,
+/// }
+///
+/// impl Component for MyComponent {
+///     type Message = ();
+///     type Properties = MyComponentProps;
+///
+///     fn view(&self) -> Html {
+///         html! {
+///             <div
+///                 class=classes!(
+///                     "my-container-class",
+///                     self.props.fill.as_some("my-fill-class"),
+///                     self.props.class.clone(),
+///                 )
+///             >
+///                 { self.props.children.clone() }
+///             </div>
+///         }
+///     }
+///
+///     // ...
+/// #   fn create(_: Self::Properties, _: ComponentLink<Self>) -> Self { unimplemented!() }
+/// #   fn update(&mut self, _: Self::Message) -> ShouldRender { unimplemented!() }
+/// #   fn change(&mut self, _: Self::Properties) -> ShouldRender { unimplemented!() }
+/// }
+/// ```
````
This isn't really an example for the `classes!` macro. Let's just keep it to a single example.
```suggestion
```

siku2 commented on yew/src/lib.rs:
```diff
@@ -98,6 +98,70 @@
 #![recursion_limit = "512"]
 extern crate self as yew;
 
+/// This macro helps you manage a list of HTML classes.
+///
+/// This macro is similar to [`classNames`](https://www.npmjs.com/package/classnames) from
+/// JavaScript. It helps you manage a set of classes more easily.
+///
+/// # Examples
```
```suggestion
/// This macro provides a convenient way to create [`Classes`].
///
/// The macro takes a list of items similar to the [`vec!`] macro and returns a [`Classes`] instance.
/// Each item can be of any type that implements `Into<Classes>` (See the implementations on [`Classes`] to learn what types can be used).
///
/// # Example
```

cecton replied:
:thinking: 

> /// This macro provides a convenient way to create [`Classes`].

I prefer my sentence. My reasoning is that `Classes` is kinda of an implementation detail. It's not like you are going to create a `Classes` object and manipulate it a lot: most of the time you just give all your classes to the macro, then pass the result to `class=` on some node, and that's it.

> /// The macro takes a list of items similar to the [`vec!`] macro and returns a [`Classes`] instance.
> /// Each item can be of any type that implements `Into<Classes>` (See the implementations on [`Classes`] to learn what types can be used).

That is more detail indeed but I like the comparison I did with classNames.

classNames is very well known among JS developers (which is probably a good part of yew's community). Giving a good example is sometimes better than a long description.

> Not perfectly happy with my suggestion but at least it goes into more detail.

I think having more details is good. But if you don't mind I think it's important to keep the comparison with classNames and the more pragmatic description "manage HTML classes".

siku2 replied:
> My reasoning is that Classes is kind of an implementation detail

And that's where we disagree. `Classes` is (and should be) part of the public API, if you disagree with that then we need to deal with that separately.
I do agree with "It's not like you are going to create a Classes object and manipulate it a lot"! I really don't like that the current implementation of `Classes` is optimized for fast manipulation rather than creation and turning into an attribute. If you want to change that, I'm all for it, but as long as the former remains true our API reference should accurately reflect that. Also, if we don't tie this to `Classes` at all we don't have a way to easily explain the return type of the macro (because users are allowed to use it outside of the `html!` macro) and what types of arguments it accepts.

> classNames is very well known among JS developers (which is probably a good part of yew's community).

I totally get it but please imagine the situation if you don't know about the library. It really comes across like "if you don't know what this is, look at the documentation of this other library written for another language". There are (surprisingly) many developers using Yew that aren't familiar with the Javascript ecosystem and as I said previously, those who are will probably recognize the similarities by themselves.

The comparison isn't even a particularly good one because they aren't 100% equal. Sure `classes!` vaguely shares the same idea of combining multiple strings into a single construct but that's about as far as it goes currently.
If you really want to, for some reason, I can just barely live with putting it on the website, but please don't put it in the docstring.

> I think having more details is good. But if you don't mind I think it's important to keep the comparison with classNames and the more pragmatic description "manage HTML classes".

The reason I replaced "manage HTML classes" is because the macro doesn't do any "managing" to speak of. It always just creates a new `Classes` instance. It's totally fine to use something like this on the website because that's supposed to serve as an introduction but the docstrings are an API reference and there is an expectation for them to be accurate and complete.

cecton replied:
:thinking: all fair points

siku2 replied:
I was so worried I was coming on too strong. Thanks for being understanding!

siku2 commented on yew/src/html/classes.rs:
```diff
@@ -0,0 +1,641 @@
+    }
+}
+
+#[cfg(test)]
+mod macro_tests {
+    extern crate self as yew;
+
+    use crate::html::AnyScope;
+    use crate::virtual_dom::{VDiff, VNode, VTag};
+    use crate::{classes, html, NodeRef};
+    use std::any::TypeId;
+    use std::rc::Rc;
+    #[cfg(feature = "std_web")]
+    use stdweb::web::{document, IElement, INode};
+    #[cfg(feature = "wasm_test")]
+    use wasm_bindgen_test::{wasm_bindgen_test as test, wasm_bindgen_test_configure};
+
+    #[cfg(feature = "wasm_test")]
+    wasm_bindgen_test_configure!(run_in_browser);
```
I'm not sure whether we should keep these tests like that. The entire reason for moving them over here was because `Classes` are no longer a direct part of the DOM. They behave like any other attribute now, so there's no need to test `classes!` in combination with `html!`.
I think we should simplify these tests so they no longer use `html!` or run in the browser. We can keep some integration tests over in `virtual_dom` just to make sure that the deprecated stuff still works for now.
What do you think?

cecton replied:
So I should:

1. add a test with `<div class=(...something...)` in virtual_dom
2. remove all the layout test (the macro is tested and there are already tests for Classes itself)

I think it sounds good

siku2 replied:
Love how GitHub shows my comment out of context in the review below :laughing:

cecton replied:
Fixed in eb7b457

siku2 reviewed (commented):
I already wanted to re-review this yesterday but you were still actively making changes so I decided to wait until today. Please don't hesitate to ping me or request a review when you're done :)

This turned out to be a bit more than I expected. Just know that the changes are already pretty good right now, I'm just a naturally nitpicky person :sweat_smile:

siku2 commented on yew-macro/src/classes/mod.rs:
```diff
@@ -0,0 +1,57 @@
+    }
+}
+
+fn class_expr_parser(input: ParseStream) -> Result<Expr> {
```
Oh and one thing we should probably do is use `.split_whitespace()` to detect whether there are multiple classes in a single string. That's what we're using in the `Classes` implementation and `split_whitespace` handles more than just `U+0020`.
This also gives us the opportunity to actually suggest a fix which is always nice:

```rust
let classes = lit.value().split_whitespace().collect::<Vec<_>>();
if classes.len() > 1 {
    let fix = classes
        .into_iter()
        .map(|class| format!("\"{}\"", class))
        .collect::<Vec<_>>()
        .join(", ");
    let msg = format!(
        "string literals must not contain more than one class (hint: use `{}`)",
        fix
    );

    Err(syn::Error::new(expr.span(), msg))
} else {
    Ok(Self::Lit(lit_str))
}
```

cecton replied:
Excellent!

cecton commented on docs/concepts/html/classes.md:
````diff
@@ -0,0 +1,89 @@
+
+```rust
+html! {
+  <div class=html!("container")></div>
````
:woman_facepalming:

cecton replied:
Oh I think I replaced globally without checking oops

cecton commented on docs/concepts/html.md:
```diff
@@ -1,12 +1,15 @@
 ---
-title: html!
+title: html! and classes!
```
Fair point!

cecton commented on docs/concepts/html/classes.md:
```diff
@@ -0,0 +1,89 @@
+
+## Classes
+
+There are a number of convenient ways to specify classes for an element:
```
Check this out 8871072

siku2 replied:
Awesome!

cecton commented:
> This turned out to be a bit more than I expected. Just know that the changes are already pretty good right now, I'm just a naturally nitpicky person sweat_smile

This is great, no worry! I think it's very important for yew to be of very good quality. I'm sorry there are a lot of stupid stuff I miss :sweat: I'll try to make an effort.

Ready for review: I have one more comment on one of your comment. Also check the doc on the side which I have improved. @siku2

siku2 commented on docs/concepts/html/classes.md:
```diff
@@ -0,0 +1,115 @@
+fact, anything that implements `Into<Classes>` can be used to push new classes
+to the set. This includes: `&'static str`, `String` and `Cow<'static, str>`,
+but also: `Vec<impl Into<Classes>>`, `&[impl Into<Classes>]`,
+`Option<impl Classes>` and `&Option<impl Classes>`. For example
```
```suggestion
`Option<impl Into<Classes>>` and `&Option<impl Into<Classes>>`. For example
```

I'm worried that this might overwhelm the user (it's a lot of "and"s in a row). What if we just use the `DOCUSAURUS_CODE_TABS` below to demonstrate the different built-in implementations?

I feel like the tab for "Interpolated" is pretty misleading anyway because it's literally just a string...

cecton replied:
Fixed in 42da556

> I'm worried that this might overwhelm the user (it's a lot of "and"s in a row). What if we just use the DOCUSAURUS_CODE_TABS below to demonstrate the different built-in implementations?

I removed the first "and". I believe it is still grammatically correct if not better but I'm not a native English speaker.

After re-reading I think it is fine as it is because:
1. the first example with the code tabs is already showing an example with the Option
2. the last example within a component shows the option but also another Classes

> I feel like the tab for "Interpolated" is pretty misleading anyway because it's literally just a string...

Yes... I think it's best to remove it entirely. 7da9ea7

cecton replied:
(I'm totally open to suggestion :grin: !)

cecton replied:
:thinking: maybe we could merge both examples into one. Use only examples withing component because `classes!` taken alone is actually pretty difficult to understand imo.

And we make every example (single, multiple, literals, options, classes...)

siku2 replied:
> thinking maybe we could merge both examples into one.

That's what I meant, yeah. That way we can explain both at the same time and it looks a bit prettier.

siku2 reviewed (commented):
Preview link so I don't have to scroll up every time: <https://yew-rs--pr1601-extract-classes-macr-edmdi2ce.web.app>

> I'm sorry there are a lot of stupid stuff I miss :sweat: I'll try to make an effort.

Don't say that, this is great stuff! And it only gets better so I can only hope to see more of your contributions in the future!

cecton commented on website/sidebars.json:
```diff
@@ -30,11 +30,12 @@
       {
         "type": "subcategory",
-        "label": "The HTML macro",
+        "label": "Yew's macros",
```
Oh right. I think I changed something else. 

Yes I changed "html! and classes!" to "HTML". :thinking:

cecton replied:
Fixed in e600f0cb8a981884e084c21bdb6fe1c7dde5b037

cecton commented on yew-macro/src/classes/mod.rs:
```diff
@@ -0,0 +1,71 @@
+    }
+}
+
+#[allow(clippy::large_enum_variant)]
+enum ClassExpr {
+    Lit(LitStr),
+    Expr(Expr),
+}
```
Would you mind sharing your knowledge on this?

I don't know much about lower level programming. I thought allocating into the heap would save memory but using more space in the stack would be faster. cc @Yozhgoor (because I gave a wrong explanation).

Though :thinking: I think sometimes things get faster because they are smaller. The speed boost is due to shifting the work load to a faster component (the CPU). It is the case for example with filesystem compression where you spend less time reading bytes from IO and more time decompressing with the CPU which overall is faster.

siku2 replied:
It's not always that straightforward. Without benchmarking these things it's really hard to tell which approach is actually faster in our case. And honestly, I don't know about most of the optimizations performed by the compiler (and especially not when it comes to LLVM). This isn't directly about speed though, Clippy complains because one variant (`Expr`) has a significantly bigger memory footprint than the others (`LitStr`). This is an issue because the size of an enum is determined by its biggest variant so in our case `ClassExpr::Lit` (which is probably the more common case) is penalized by a ton padding which in return can reduce performance.

If you look at what's contained in an `Expr` value you'll find that most of it is already on the heap anyway (either in the form of `Vec<_>` or `Box<_>`) so I don't think the overhead of the allocation causes a noticeable speed difference here. Parsing the `proc_macro2::TokenStream` into an `Expr` is probably the dominating factor for performance here anyway.

cecton replied:
What I explained initially is that this warns that the enum is taking too much memory because of the difference between the size of the variants is big. So putting the big variant into a Box saves memory when the smaller variant is used. But since we are in a macro, I suppose it's less of a problem. It's not like we are going to instantiate millions of classes during the compilation. That is why I think in this case we can safely ignore it.

But yeah it wasn't hard to fix either :confounded:

cecton commented on yew-macro/src/classes/mod.rs:
```diff
@@ -0,0 +1,71 @@
+
+impl Parse for Classes {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        input.parse_terminated(|x| ClassExpr::parse(x)).map(Self)
```
This is fine.

Though I must say it's an optimization I find tedious and I often find clippy annoying for reporting something like that. Also because I find the closure more readable

cecton replied:
Fixed in 3482bf8

cecton commented on yew-macro/src/html_tree/html_element.rs:
```diff
@@ -234,17 +234,35 @@ impl ToTokens for HtmlElement {
+                };
+
+                Some(quote! {
+                    {
```
I just checked and no it wasn't intended :grin: I just pressed `D` in VIM to quickly remove the span when it was quote_spanned but I didn't realize I had 2 blocks now.

cecton replied:
Fixed in f8e1e5d

cecton commented on yew-macro/src/classes/mod.rs:
```diff
@@ -0,0 +1,71 @@
+                }
+            }
+            expr => Ok(Self::Expr(expr)),
+        }
```
Fixed in 2dfe09a

cecton commented on yew-macro/src/html_tree/html_element.rs:
```diff
@@ -254,19 +272,21 @@ impl ToTokens for HtmlElement {
                         let sr = lit.stringify();
                         Some(quote! {
-                            #vtag.__macro_push_attribute("class", #sr);
+                             #vtag.__macro_push_attribute("class", #sr);
```
Fixed in 002103c

cecton commented on yew-macro/src/html_tree/html_element.rs:
```diff
@@ -367,6 +387,13 @@ impl ToTokens for HtmlElement {
+                        Please use the macro `classes!` instead.",
+                )]
+                fn deprecated_use_of_class() {}
+
```
Fixed in f8e1e5d

cecton commented on yew-macro/tests/macro_test.rs:
```diff
@@ -2,7 +2,7 @@ use yew::html;
 #[allow(dead_code)]
 #[rustversion::attr(stable(1.45), test)]
-fn tests() {
+fn macros() {
```
I noticed because I had a hard time targeting my new tests XD

Fixed in ee491cd

cecton commented on yew-macro/tests/classes_macro/classes-fail.stderr:
```diff
@@ -0,0 +1,107 @@
+ --> $DIR/classes-fail.rs:4:5
+  |
+4 |     classes!(42);
+  |     ^^^^^^^^^^^^^ the trait `std::convert::From<{integer}>` is not implemented for `yew::html::classes::Classes`
```
Fixed in e1a6480

cecton commented on yew/src/html/classes.rs:
```diff
@@ -0,0 +1,647 @@
+    iter::FromIterator,
+};
+
+/// A set of classes.
```
Fixed in c91407d

I couldn't find a way to make it clickable. I tried:

```
[`classes!`]
[`crate::classes!`]
[`self::classes!`]
[`yewclasses!`]
```

Nothing worked

siku2 replied:
Ah, you can't add the exclamation mark (for technical reasons). I would write it like this: [\`classes!`][yew::classes]
That way it still shows up as `classes!`.

*Oh and just to make sure, intra doc links still require nightly Rustdoc for the moment (I think). The feature was stabilized a few weeks ago but no new version has been released yet.*

siku2 replied:
Ah never mind, I was wrong. Linking to `yew::classes!` works on my end. Might be that you aren't using the nightly version locally?

Either way, I'd still write it as: [`classes!][yew::classes!] (can't use just [\`classes!`] because it's not in scope)

cecton replied:
Ah yes I don't use nightly. Ok it's fixed and working (in nightly) 76db907

cecton commented:
Ready for review cc @siku2

siku2 commented on yew-macro/tests/html_macro_test.rs:
```diff
@@ -0,0 +1,46 @@
+fn html_macro() {
+    let t = trybuild::TestCases::new();
+
+    t.pass("tests/html_macro/html-block-pass.rs");
+    t.compile_fail("tests/html_macro/html-block-fail.rs");
+
+    t.pass("tests/html_macro/html-component-pass.rs");
+    t.compile_fail("tests/html_macro/html-component-fail.rs");
+    t.compile_fail("tests/html_macro/html-component-fail-unimplemented.rs");
+
+    t.pass("tests/html_macro/html-iterable-pass.rs");
+    t.compile_fail("tests/html_macro/html-iterable-fail.rs");
+
+    t.pass("tests/html_macro/html-list-pass.rs");
+    t.compile_fail("tests/html_macro/html-list-fail.rs");
+
+    t.pass("tests/html_macro/html-node-pass.rs");
+    t.compile_fail("tests/html_macro/html-node-fail.rs");
+
+    t.pass("tests/html_macro/html-element-pass.rs");
+    t.compile_fail("tests/html_macro/html-element-fail.rs");
```
while we're at it, let's switch to the wildcard approach for these:
```suggestion
    t.pass("tests/html_macro/*-pass.rs");
    t.compile_fail("tests/html_macro/*-fail.rs");
```

The only problem is the `html-component-fail-unimplemented.rs` fail, let's just rename that one to `html-component-unimplemented-fail.rs` so it matches the pattern.

siku2 commented on yew/src/html/classes.rs:
```diff
@@ -0,0 +1,242 @@
+    iter::FromIterator,
+};
+
+/// A set of classes. The use of `classes!` is preferred.
```
```suggestion
/// A set of classes.
/// 
/// Using [`classes!`][yew::classes!] for initialization is preferred.
```

I like to draw parallels between `Vec/vec!` and `Classes/classes!` and well... Let's just say our documentation doesn't hold up in comparison :smile:
Well anyway, I'm hoping some day a magic documentation guru shows up and writes the perfect documentation for us :pray:

siku2 reviewed (commented):
~~GitHub kind of broke because a new commit was added while I was reviewing this. Let's hope it still turned out okay.~~
Looks like it survived. This should be the last review for the code itself :tada: (unless the final re-review surfaces some issue I missed, of course)

siku2 commented:
Okay that's not really how I imagined it. The code examples are way too verbose now and it's hard to tell what to look at.

cecton commented:
> Okay that's not really how I imagined it. The code examples are way too verbose now and it's hard to tell what to look at.

It can be cleaned-up a bit by removing the Boolinator import where it's not used and the struct with the component itself

cecton commented:
@siku2 do you have any idea for this? I an revert to a previous state if you prefer. Or do just as I said (cleanup things a bit)

siku2 commented:
> @siku2 do you have any idea for this? I an revert to a previous state if you prefer. Or do just as I said (cleanup things a bit)

Ah sorry, I thought you were still playing around with it to see what's best. I honestly preferred the previous version because the code in the tabs should be as concise as possible (otherwise it distracts from the point of each tab). There was also possibly a misunderstanding because I was thinking about combining the list of `Into<Classes>` implementations into these tabs (so instead of listing all types which can be turned into `Classes` at the top, we use the tabs to do it). I feel like that's easier to follow than this:
> This includes: `&'static str`, `String` and `Cow<'static, str>`,
but also: `Vec<impl Into<Classes>>`, `&[impl Into<Classes>]`,
`Option<impl Classes>` and `&Option<impl Classes>`.

cecton commented:
Now I understand why they want to rewrite the React's doc :sweat_smile: Apparently it gets messy because a lot of people wrote with their own style and at some point it isn't consistent.

So... I reverted the latest change and instead I added an example tab for String and another one for Array ref. I'm not sure it would make much sense with Cow... I think the Cow is mostly here to handle the fact that we mix String and literals, I don't see a use case.

Let me know what you think with this latest changes. Feel also free to correct by yourself, I really don't mind at all.

siku2 commented:
> Now I understand why they want to rewrite the React's doc sweat_smile Apparently it gets messy because a lot of people wrote with their own style and at some point it isn't consistent.

Haha yeah, Yew's documentation is a mess of different styles.

> So... I reverted the latest change and instead I added an example tab for String and another one for Array ref. I'm not sure it would make much sense with Cow... I think the Cow is mostly here to handle the fact that we mix String and literals, I don't see a use case.
> 
> Let me know what you think with this latest changes. Feel also free to correct by yourself, I really don't mind at all.

Looks good enough :)
I'll try to squeeze in another review tomorrow just to make sure I didn't miss anything important but otherwise I think we're good to go
